src/components/external-source/ExternalSourceManager.svelte

```diff
@@ -0,0 +1,1432 @@
+  let canvasMouseOverEvent: MouseEvent | undefined = undefined;
+  let removeDPRChangeListener: (() => void) | null = null;
+  let eventTooltip: Tooltip;
+  let mouseOver: MouseOver | null;
+
+  // We want to parse a file selected for upload.
+  let files: FileList | undefined;
+  let file: File | undefined;
+  let parsed: ExternalSourceJson | undefined;
+
+  // External event type creation variables
+  let externalEventTypeInsertInput: ExternalEventTypeInsertInput;
+  let externalEventsCreated: ExternalEventInsertInput[] = [];
+
+  // For filtering purposes (modelled after TimelineEditorLayerFilter):
+  let filterMenu: Menu;
+  let input: HTMLInputElement;
+  let filterString: string = '';
+  let filteredValues: ExternalSourceType[] = [];
+  let selectedFilters: ExternalSourceType[] = [{ name: '' }];
+  let menuTitle: string = '';
+  let filteredExternalSources: ExternalSourceSlim[] = [];
+
+  // External source + derivation group creation variables
+  let sourceInsert: ExternalSourceInsertInput;
+  let sourceTypeInsert: ExternalSourceTypeInsertInput;
+  let derivationGroupInsert: DerivationGroupInsertInput;
+  let selectedSourceLinkedDerivationGroupsPlans: PlanDerivationGroup[] = [];
+
+  // Permissions
+  let hasDeletePermission: boolean = false;
+  let hasCreatePermission: boolean = false;
+
+  // There was a strange issue where when:
+  //   - you select a source,
+  //   - select an event in timeline,
+  //   - go to table,
+  //   - select a different source,
+  //   - select an event,
+  //   - then go back to timeline.
+  //  The event autodeselected. Some prints led to the discovery that an onMouseDown gets fired, somewhere on the canvas, immediately after selection.
+  //  As this mouseDown only occurs after the table switches back to a timeline, a simple boolean check was added to remedy this, saying to ignore
+  //    the mouseDown occurring right after a switch from a table to a timeline.
+  let mouseDownAfterTable: boolean = false;
+
+  let isDerivationGroupFieldDisabled: boolean = true;
+
+  let gridRowSizes: string = '1fr 3px 0fr';
+
+  // Clear all error stores when a source is selected as they will not be shown
+  $: if (selectedSource !== null) {
+    createExternalSourceError.set(null);
+    createExternalSourceTypeError.set(null);
+    createDerivationGroupError.set(null);
+    parsingError.set(null);
+  }
+
+  $: selectedSourceId = selectedSource ? getRowIdExternalSource(selectedSource.pkey) : null;
+
+  // File parse logic
+  $: if (files) {
+    // Safeguard against infinitely executing parse logic
+    if (file !== files[0]) {
+      createExternalSourceError.set(null);
+      createExternalSourceTypeError.set(null);
+      createDerivationGroupError.set(null);
+      parsingError.set(null);
+      isDerivationGroupFieldDisabled = true;
+
+      file = files[0];
+      const fileText = file.text();
+      fileText.then(async text => {
+        try {
+          parsed = JSON.parse(await text);
```

Comment:
Based on the limit, should we flash a modal or a warning somewhere indicating that a file is too big? Or is there other desired behavior?

Reply:
The string limit is pretty big depending on the system/browser, so it might be a little hard to determine if it's too big. Chrome's limit is a string of 512mb on a 32-bit system. Do you think these event files can be that big? We do have a way of streaming in a plan JSON file into a parser if you wanted to use that as an [example](https://github.com/NASA-AMMOS/aerie-ui/blob/develop/src/routes/plans/%2Bpage.svelte#L645).

Reply:
Went ahead and implemented file streaming for these files same as the Plan JSONs use - I don't *think* we should ever reach a cap because even with our 10,000+ event test JSONs we only reach ~30MB, but definitely better safe than sorry!